File: javascript/Easy/containsDuplicate.js
const nums1 = [1,2,3,1]
// Output: true

const nums2 = [1,2,3,4]
// Output: false

const nums3 = [1,1,1,3,3,4,3,2,4,2]
// Output: true



const containsDuplicate = function(nums) {
  let currentIndex = 0;
  for (let i = 0; i < nums.length; i++) {
  
      // verifies that the matching duplicate is not the same value and then returns true
      if (nums[currentIndex] === nums[i] && i !== currentIndex) {
          return true;
      }
      
      // verifies that loop has completed, increments index, starts loop again
      if (i === nums.length - 1 && currentIndex !== nums.length - 1) {
          currentIndex++;
          i = 0;
      }
      
      // verifies that all combinations have been tried, stops loop
      if (i === nums.length - 1 && currentIndex === nums.length - 1) {
          return false;
      }
  }
};

console.log(containsDuplicate(nums1));
console.log(containsDuplicate(nums2));
console.log(containsDuplicate(nums3));

// PASSED 

File: C++/Easy/containsDuplicate.cpp
#include <iostream>
#include <vector>
#include <unordered_set>

using namespace std;

class Solution {
  public:
    bool containsDuplicate(vector<int> & nums) {
      unordered_set<int> s;

      for (int i = 0; i < nums.size(); i++) {
        if (s.find(nums[i]) != s.end()) {
          return true;
        };
        s.insert(nums[i]);
      };

      return false;
    };
};

int main() {
  Solution sol;

  vector<int> nums1 = {1, 2, 3, 1};
  vector<int> nums2 = {1, 2, 3, 4};
  vector<int> nums3 = {1, 1, 1, 3, 3, 4, 3, 2, 4, 2};

  bool result1 = sol.containsDuplicate(nums1);
  bool result2 = sol.containsDuplicate(nums2);
  bool result3 = sol.containsDuplicate(nums3);

  cout << "Contains duplicate: " << (result1 ? "true" : "false") << endl;
  cout << "Contains duplicate: " << (result2 ? "true" : "false") << endl;
  cout << "Contains duplicate: " << (result3 ? "true" : "false") << endl;

  return 0;
};

// run the following in the terminal
/* g++ -std=c++11 containsDuplicate.cpp -o  containsDuplicate */
// then
/* ./containsDuplicate */